wavesexchange_repos/src/circuit_breaker/mod.rs

```diff
@@ -0,0 +1,249 @@
+    data_source: Arc<S>,
+    err_count: usize,
+    first_err_ts: Option<Instant>,
+}
+
+impl<S: FallibleDataSource> CBState<S> {
+    fn inc(&mut self) {
+        self.err_count += 1;
+    }
+
+    fn reset(&mut self) {
+        self.err_count = 0;
+        self.first_err_ts = None;
+    }
+
+    fn reinit(&mut self, src: S) {
+        self.data_source = Arc::new(src);
+    }
+}
+
+pub struct CircuitBreakerBuilder<S: FallibleDataSource> {
+    max_timespan: Option<Duration>,
+    max_err_count_per_timespan: Option<usize>,
+    init_fn: Option<Box<dyn DataSrcInitFn<S>>>,
+}
+
+impl<S: FallibleDataSource> CircuitBreakerBuilder<S> {
+    pub fn new() -> CircuitBreakerBuilder<S> {
+        CircuitBreakerBuilder {
+            max_timespan: None,
+            max_err_count_per_timespan: None,
+            init_fn: None,
+        }
+    }
+
+    pub fn with_max_timespan(mut self, ts: Duration) -> CircuitBreakerBuilder<S> {
+        self.max_timespan = Some(ts);
+        self
+    }
+
+    pub fn with_max_err_count_per_timespan(mut self, count: usize) -> CircuitBreakerBuilder<S> {
+        self.max_err_count_per_timespan = Some(count);
+        self
+    }
+
+    pub fn with_init_fn(mut self, f: impl DataSrcInitFn<S>) -> CircuitBreakerBuilder<S> {
+        self.init_fn = Some(Box::new(f));
+        self
+    }
+
+    pub fn build(self) -> Result<CircuitBreaker<S>, S::Error> {
+        // probably there is a better way to force use all with_* methods on builder
+        if self.max_err_count_per_timespan.is_none() {
+            panic!("max_err_count_per_timespan is not set");
+        }
+
+        if self.max_timespan.is_none() {
+            panic!("max_timespan is not set");
+        }
+
+        if self.init_fn.is_none() {
+            panic!("init_fn is not set");
+        }
+
+        let init_fn = self.init_fn.unwrap();
+
+        Ok(CircuitBreaker {
+            state: RwLock::new(CBState {
+                data_source: Arc::new(init_fn()?),
+                err_count: 0,
+                first_err_ts: None,
+            }),
+            max_timespan: self.max_timespan.unwrap(),
+            max_err_count_per_timespan: self.max_err_count_per_timespan.unwrap(),
+            init_fn,
+        })
+    }
+}
+
+impl<S: FallibleDataSource> CircuitBreaker<S> {
+    pub fn builder() -> CircuitBreakerBuilder<S> {
+        CircuitBreakerBuilder::new()
+    }
+
+    pub fn builder_from_cfg(cfg: &Config) -> CircuitBreakerBuilder<S> {
+        Self::builder()
+            .with_max_err_count_per_timespan(cfg.max_err_count_per_timespan)
+            .with_max_timespan(cfg.max_timespan)
+    }
+
+    pub async fn query<T, F, Fut>(&self, query_fn: F) -> Result<T, S::Error>
+    where
+        F: FnOnce(Arc<S>) -> Fut,
+        Fut: Future<Output = Result<T, S::Error>>,
+    {
+        let state_read_lock = self.state.read().await;
+        let result = query_fn(state_read_lock.data_source.clone()).await;
+
+        drop(state_read_lock);
```

Comment:
This looks silly - the thread-safe dasta structure such as Arc was put under mutex, even async mutex, and now we need to lock that mutex just to clone the Arc.

Reply:
CB is designed to interfere client's code as less as possible. RwLock is used here to provide interior state mutability, without it, the end client will suffer while refactoring `&self` to `&mut self`. 
`Arc` is used to pass repo to query function. I'm not using `&` due to lifetime problems (`query_fn` is async)